Mathlib/RingTheory/MvPowerSeries/Topology.lean

```diff
@@ -0,0 +1,311 @@
+
+/-- MvPowerSeries on a T2Space form a T2Space -/
+@[scoped instance] theorem t2Space [T2Space α] : T2Space (MvPowerSeries σ α) where
+  t2 x y h := by
+    obtain ⟨d, h⟩ := Function.ne_iff.mp h
+    obtain ⟨u, v, ⟨hu, hv, hx, hy, huv⟩⟩ := t2_separation h
+    exact ⟨(fun x => x d) ⁻¹' u, (fun x => x d) ⁻¹' v,
+      IsOpen.preimage (continuous_component σ α d) hu,
+      IsOpen.preimage (continuous_component σ α d) hv, hx, hy,
+      Disjoint.preimage _ huv⟩
+
+end WithPiTopology
+
+end Topological
+
+section Uniform
+
+namespace WithPiUniformity
+
+open WithPiTopology
+
+variable [UniformSpace α]
+
+/-- The componentwise uniformity on MvPowerSeries -/
+scoped instance uniformSpace : UniformSpace (MvPowerSeries σ α) :=
+  Pi.uniformSpace fun _ : σ →₀ ℕ => α
+
+/-- Components are uniformly continuous -/
+theorem uniformContinuous_component :
+    ∀ d : σ →₀ ℕ, UniformContinuous fun a : MvPowerSeries σ α => a d :=
+  uniformContinuous_pi.mp uniformContinuous_id
+
+/-- The uniform_add_group structure on MvPowerSeries of a uniform_add_group -/
+@[scoped instance] theorem uniformAddGroup [AddGroup α] [UniformAddGroup α] :
+    UniformAddGroup (MvPowerSeries σ α) where
+  uniformContinuous_sub := uniformContinuous_pi.mpr fun _ => UniformContinuous.comp
+    uniformContinuous_sub
+    (UniformContinuous.prod_mk
+      (UniformContinuous.comp (uniformContinuous_component _ _ _) uniformContinuous_fst)
+      (UniformContinuous.comp (uniformContinuous_component _ _ _) uniformContinuous_snd))
+
+/-- Completeness of the uniform structure on MvPowerSeries -/
+@[scoped instance] theorem completeSpace [AddGroup α] [CompleteSpace α] :
+    CompleteSpace (MvPowerSeries σ α) where
+  complete := by
+    intro f hf
+    suffices ∀ d, ∃ x, (f.map fun a => a d) ≤ nhds x by
+      use fun d => (this d).choose
+      rw [nhds_pi, Filter.le_pi]
+      exact fun d => (this d).choose_spec
+    intro d
+    use lim (f.map fun a => a d)
+    exact (Cauchy.map hf (uniformContinuous_component σ α d)).le_nhds_lim
+
+/-- Separation of the uniform structure on MvPowerSeries -/
+@[scoped instance] theorem t0Space [T0Space α] : T0Space (MvPowerSeries σ α) := by
+  suffices T2Space (MvPowerSeries σ α) by infer_instance
+  exact WithPiTopology.t2Space σ α
+
+/-- The ring of multivariate power series is a uniform topological ring -/
+@[scoped instance] theorem uniform_topologicalRing [Ring α] [UniformAddGroup α] [TopologicalRing α] :
+    TopologicalRing (MvPowerSeries σ α) :=
+  { uniformAddGroup σ α with
+    continuous_add :=  (@uniformContinuous_add _ _ _ (uniformAddGroup σ α)).continuous
+    continuous_mul := continuous_pi fun _ => continuous_finset_sum _ fun i _ => Continuous.comp
+      continuous_mul (Continuous.prod_mk (Continuous.fst' (continuous_component σ α i.fst))
+        (Continuous.snd' (continuous_component σ α i.snd)))
+    continuous_neg := (@uniformContinuous_neg _ _ _ (uniformAddGroup σ α)).continuous
+    }
+
+end WithPiUniformity
+
+end Uniform
+
```

Comment:
```suggestion
```

Reply:
Here I want to keep some variables and the `open…`